### PR TITLE
adding ledgerbackend datastore txmeta as a data source

### PR DESCRIPTION
**Pull request by chowbao**
Integrate cdp changes to ingest from datastore txmeta files
* Added new option to read/create ledgerbackend using the ledgerbackend.datastore to interface with the txmeta files from gcs
* Controlled by two new parameters
  * captive-core - will use captive core; datastore gcs is the default for stellar-etl now
  * datastore-url - the url to the datastore (gcs bucket)
* Most export_* could use the LedgerCloseMeta to transform/process the data needed
* assets and ledgers which originally pulled from history archives had a custom xdr format that needed to be manually constructed from the LedgerCloseMeta
  * TODO: In the future could consider removing this dependency all together and use straight LedgerCloseMeta instead (different PR)
* TODO: Probably should delete all the unused from_genesis export_*s (different PR)
* TODO: All the parameters are kinda a mess; could be refactored (different PR)
* TODO: The export_* commands could also be refactored (different PR)

### CODE REVIEW

File: cmd/export_ledger_entry_changes.go [hunk 1]
```diff
@@ -46,7 +50,7 @@ be exported.`,
 		// If none of the export flags are set, then we assume that everything should be exported
 		allFalse := true
 		for _, value := range exports {
-			if true == value {
+			if value {
```
Comment (sydneynotthecity) [inline cmd/export_ledger_entry_changes.go:53]: lol

File: internal/input/assets_history_archive.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"context"
+
+	"github.com/stellar/stellar-etl/internal/transform"
+	"github.com/stellar/stellar-etl/internal/utils"
+
+	"github.com/stellar/go/xdr"
+)
+
+// GetPaymentOperations returns a slice of payment operations that can include new assets from the ledgers in the provided range (inclusive on both ends)
+func GetPaymentOperationsHistoryArchive(start, end uint32, limit int64, isTest bool, isFuture bool) ([]AssetTransformInput, error) {
```
Comment (sydneynotthecity) [inline internal/input/assets_history_archive.go:13]: even though the backend is different, might be nice to maintain parity between `GetPaymentOperations` and `GetPaymentOperationsHistoryArchive`. ie, maybe keep the input parameters for the methods consistent.

File: internal/input/ledgers_history_archive.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"context"
+
+	"github.com/stellar/stellar-etl/internal/utils"
+
+	"github.com/stellar/go/historyarchive"
+)
+
+// GetLedgers returns a slice of ledger close metas for the ledgers in the provided range (inclusive on both ends)
+func GetLedgersHistoryArchive(start, end uint32, limit int64, isTest bool, isFuturenet bool) ([]historyarchive.Ledger, error) {
```
Comment (sydneynotthecity) [inline internal/input/ledgers_history_archive.go:12]: same feedback as `GetPaymentOperation`

File: internal/input/operations.go [hunk 4]
```diff
@@ -40,15 +39,14 @@ func GetOperations(start, end uint32, limit int64, env utils.EnvironmentDetails)
...
-		changeReader, err := ingest.NewLedgerChangeReader(ctx, backend, env.NetworkPassphrase, seq)
+		ledgerCloseMeta, err := backend.GetLedger(ctx, seq)
 		if err != nil {
-			return []OperationTransformInput{}, err
+			return []OperationTransformInput{}, fmt.Errorf("error getting ledger seq %d from the backend: %v", seq, err)
 		}
-		txReader := changeReader.LedgerTransactionReader
 
-		ledgerCloseMeta, err := backend.GetLedger(ctx, seq)
+		txReader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(env.NetworkPassphrase, ledgerCloseMeta)
```
Comment (sydneynotthecity) [inline internal/input/operations.go:47]: nice, this is much cleaner

File: internal/utils/main.go [hunk 5]
```diff
@@ -713,3 +743,21 @@ func LedgerEntryToLedgerKeyHash(ledgerEntry xdr.LedgerEntry) string {
...
+func CreateLedgerBackend(ctx context.Context, useCaptiveCore bool, env EnvironmentDetails) (ledgerbackend.LedgerBackend, error) {
+	if useCaptiveCore {
+		backend, err := env.CreateCaptiveCoreBackend()
+		if err != nil {
+			return nil, err
+		}
+		return backend, nil
+	}
+
+	backend, err := ledgerbackend.NewCloudStorageBackend(ctx, env.StorageURL)
```
Comment (sydneynotthecity) [inline internal/utils/main.go:758]: is this method going to be agnostic to cloud provider? or will each cloud provider have its own backend?

File: cmd/export_assets.go [hunk 6]
```diff
@@ -16,14 +16,22 @@ var assetsCmd = &cobra.Command{
...
 		outFile := mustOutFile(path)
 
-		paymentOps, err := input.GetPaymentOperations(startNum, endNum, limit, isTest, isFuture)
+		var paymentOps []input.AssetTransformInput
+		var err error
+
+		if useCaptiveCore {
+			paymentOps, err = input.GetPaymentOperationsHistoryArchive(startNum, endNum, limit, isTest, isFuture)
+		} else {
+			paymentOps, err = input.GetPaymentOperations(startNum, endNum, limit, env, useCaptiveCore)
```
Comment (sydneynotthecity) [inline cmd/export_assets.go:33]: I like your idea of eventually refactoring this History Archives method to use lcm directly. I think it's cleaner and less confusing

**Review by sydneynotthecity — COMMENTED** [review]
Great start for ETL refactor. Small nit to get the unit test passing for `operation_trace_code` again.

I agree with all of your TODOs listed and think we should document them in tickets to be working in an upcoming sprint. 

Did you test the code using the `captive_core` flag to confirm that the code is backwards compatible?

File: internal/utils/main.go (hunk 5 above)
Comment (chowbao) [reply internal/utils/main.go:758]: Yes it is cloud provider agnostic. It's just only the GCS datastore is defined right now

**Comment by chowbao** [comment]
Before merging:
* Wait for ledgerbackend to be merged and use a non-branch go repo https://github.com/stellar/go/pull/5260
* Create and point to a prod gcs bucket for txmeta instead of the ledgerexporter-test in dev

File: internal/utils/main.go [hunk 7]
```diff
@@ -714,10 +744,99 @@ func LedgerEntryToLedgerKeyHash(ledgerEntry xdr.LedgerEntry) string {
...
 	ledgerKeyByte, _ := ledgerKey.MarshalBinary()
 	hashedLedgerKeyByte := hash.Hash(ledgerKeyByte)
 	ledgerKeyHash := hex.EncodeToString(hashedLedgerKeyByte[:])
 
 	return ledgerKeyHash
 }
+
+// AccountSignersChanged returns true if account signers have changed.
+// Notice: this will return true on master key changes too!
+func AccountSignersChanged(c ingest.Change) bool {
```
Comment (chowbao) [inline internal/utils/main.go:775]: This function was removed from the ingest library. Needed for account signers

**Review by sydneynotthecity — APPROVED** [review]
LGTM

**Pull request merged**